pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java

```diff
@@ -174,9 +174,14 @@ public void init(PinotConfiguration config, InstanceDataManager instanceDataMana
     String joinOverflowModeStr = config.getProperty(CommonConstants.MultiStageQueryRunner.KEY_OF_JOIN_OVERFLOW_MODE);
     _joinOverflowMode = joinOverflowModeStr != null ? JoinOverFlowMode.valueOf(joinOverflowModeStr) : null;
 
-    _executorService =
-        ExecutorServiceUtils.create(config, Server.MULTISTAGE_EXECUTOR_CONFIG_PREFIX, "query-runner-on-" + port,
-            Server.DEFAULT_MULTISTAGE_EXECUTOR_TYPE);
+    int hardLimit = ExecutorServiceUtils.getMultiStageExecutorHardLimit(config);
+    String executorType = Server.DEFAULT_MULTISTAGE_EXECUTOR_TYPE;
+    if (hardLimit > 0) {
+      executorType = "hardlimit";
+    }
+    _executorService = ExecutorServiceUtils.create(config, Server.MULTISTAGE_EXECUTOR_CONFIG_PREFIX,
+          "query-runner-on-" + port, executorType);
```

Comment:
I don't think you need to use the ExecutorServiceUtils.create here. You always want to use a specific executor service, so use it directly.

Reply:
Yep, definitely I overstepped trying to use the Provider API for that.

a974159d74a63116ef234b0617ca60aaac5b042f